rfcs/013-node-types.md

```diff
@@ -0,0 +1,145 @@
+---
+name: Node Types
+about: Introduce heterogeneous node types to CouchDB 4
+title: 'Node Types'
+labels: rfc, discussion
+assignees: ''
+
+---
+
+# Introduction
+
+This RFC proposes the ability to have different node types in CouchDB 4+. This
+would improve performance and allow for a more efficient use of resources.
+
+## Abstract
+
+Previously, in CouchDB 2 and 3, cluster functionality was uniformly distributed
+amongst the nodes. Any node could accept HTTP requests, run replication jobs
+and build secondary indices. With the FDB-based topology, CRUD operations have
+lower resource needs and so it could be useful to have a heterogeneous
+topology, where for example, CRUD operations run on lower capacity nodes, and a
+few higher capacity nodes handle replication or indexing jobs.
+
+
+## Requirements Language
+
+The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT",
+"SHOULD", "SHOULD NOT", "RECOMMENDED",  "MAY", and "OPTIONAL" in this
+document are to be interpreted as described in
+[RFC 2119](https://www.rfc-editor.org/rfc/rfc2119.txt).
+
+## Terminology
+
+*node type* : A label used to designate a subset of CouchDB functionality.
+
+---
+
+# Detailed Description
+
+## Node Types
+
+A node type is a string description of a some internal CouchDB
+functionality. These are the initially defined node types:
+
+ * "api_frontend" : Indicates this node can accept HTTP API requests.
+ * "view_indexing" : Indicates this node can build map/reduce view indices.
+ * "search_indexing": Indicates this node can build search indices.
+ * "replication" : Indicates this node can run replication jobs.
+
+Users can configure CouchDB nodes with any combination of those types.
+
+## Configuration
+
+Configuration MAY be specified in the standard `.ini` files or via environment
+variables. Environment variables have a higher priority, and override the
+`.ini` based settings. By default, if not configured either via `.ini` files or
+environment variable, a node type is considered enabled i.e. the value defaults
+to `true`.
```

Comment:
We’ve spent some good amount of work in the past to remove .ini-based configuration that might have influence about which code is loaded and executed for security reasons. I would assume, if the replicator role is not enabled, the entire module wouldn’t be loaded and started? If that’s the case, I’d be in favour of only configuring this via env vars and .app files rather than the .ini system.

Reply:
In the case background jobs, which use the `couch_jobs` framework, the application will probably be running on all nodes but some parts of it will be selectively disabled. 

For example, for replication, the frontend nodes might still need to call something like `couch_replicator:scheduler_jobs()`, or `couch_replicator:add_job()`. The backend will have the application running but only with have parts that do the `couch_jobs:accept(<<"replication">>, ...)` calls.

However it is a fair point that configuration will affect which top level supervisor children will start and stop, and which modules might get loaded. So I agree that it makes sense to replace the `.ini` configuration with application environment config options.

How does something like this look?

va.args:
```
-fabric node_types '[{api_frontend, true}, {view_indexing, false}]'
```

Reply:
lgtm